subprojects/core/src/main/java/org/gradle/api/internal/project/LifecycleAwareProject.java

```diff
@@ -16,39 +16,53 @@
 
 package org.gradle.api.internal.project;
 
+import org.gradle.api.IsolatedAction;
 import org.gradle.internal.Cast;
 import org.gradle.internal.metaobject.DynamicInvokeResult;
 import org.gradle.internal.metaobject.DynamicObject;
 import org.gradle.internal.metaobject.DynamicObjectUtil;
 import org.gradle.internal.reflect.Instantiator;
+import org.gradle.invocation.EagerLifecycleExecutor;
 
 import javax.annotation.Nullable;
 
 /**
  * Wrapper for {@link ProjectInternal} that has been accessed across projects, even in vintage mode.
+ * <p>
+ * When mutable state accessed - executes previously registered {@link org.gradle.api.invocation.GradleLifecycle#beforeProject(IsolatedAction)}
+ * actions <b>eagerly</b>.
  */
 public class LifecycleAwareProject extends MutableStateAccessAwareProject {
 
-    public static ProjectInternal from(ProjectInternal target, ProjectInternal referrer, Instantiator instantiator) {
+    public static ProjectInternal from(
+        ProjectInternal target,
+        ProjectInternal referrer,
+        EagerLifecycleExecutor eagerLifecycleExecutor,
+        Instantiator instantiator
+    ) {
         return MutableStateAccessAwareProject.wrap(
             target,
             referrer,
-            project -> instantiator.newInstance(LifecycleAwareProject.class, target, referrer)
+            project -> instantiator.newInstance(LifecycleAwareProject.class, target, referrer, eagerLifecycleExecutor)
         );
     }
 
-    public LifecycleAwareProject(ProjectInternal delegate, ProjectInternal referrer) {
+    private final EagerLifecycleExecutor eagerLifecycleExecutor;
+
+    public LifecycleAwareProject(ProjectInternal delegate, ProjectInternal referrer, EagerLifecycleExecutor eagerLifecycleExecutor) {
         super(delegate, referrer);
+        this.eagerLifecycleExecutor = eagerLifecycleExecutor;
     }
 
     @Override
     protected void onMutableStateAccess(String what) {
-        // no-op for now
+        eagerLifecycleExecutor.executeBeforeProjectFor(delegate);
     }
 
     @Override
     protected Object propertyMissing(String name) {
         onMutableStateAccess("getProperty");
+        // TODO just delegate.getProperty?
```

Comment:
❌

Reply:
I want to keep it, to simplify methods implementation in future